### PR TITLE
Add azure openai deployment connector

### PR DESCRIPTION
**Pull request by isuruRuhu**
# Description

Add a connector for azure openAI model deployment

Fixes https://github.com/wso2-enterprise/choreo/issues/19293

One line release note: 
- Add a connector for azure openAI model deployment

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Configuration**:
* Ballerina Version: 2201.4.1
* Operating System: macOS Ventura (version 13.0.1)
* Java SDK: openjdk version "17.0.6"

# Checklist:

### Changes to OpenAPI definition

- [x] Modified `info` section of the OpenAPI definition file - [Example](https://github.com/ballerina-platform/ballerina-extended-library/discussions/74)

    - `description` - what the connector is about 
    - `x-ballerina-init-description` OpenAPI extension - connector initialization details 
    - `x-ballerina-display` OpenAPI extension - connector name and path to the connector icon 

- [x] Added proper description each API Key, if API Key authentication is defined in OpenAPI definition file 

### Changes to connector module

- [x] Added license header at the top of each .bal file. 
- [x] Added Package.md as per the guide [here](https://github.com/ballerina-platform/ballerina-extended-library/discussions/77)
- [x] Added Module.md for the module of the  connector as per the guide [here](https://github.com/ballerina-platform/ballerina-extended-library/discussions/78)
- [x] Added connector icon to the root of the connector. Icon needs to be a `png` of 200x200 px size with name `icon.png`
- [x] Added Ballerina.toml file with following information. For keywords refer to [guide](https://github.com/ballerina-platform/ballerina-extended-library/discussions/72)

### Verifying connector module

- [x] Compiled the connector successfully with the targeted ballerina version. 
- [ ] Conducted smoke tests on the connector (Recommended when OpenAPI definition is obtain from an unofficial source)

### Security checks
 - [x] Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? 
 - [x] Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets?

### CODE REVIEW

File: openapi/azure.openai.deployment/Module.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
+## Overview
+This is a generated connector from [Azure OpenAI Deployments API](https://learn.microsoft.com/en-us/rest/api/cognitiveservices/azureopenaistable/deployments/) OpenAPI specification.
+
+The Azure OpenAI Deployments REST API endpoints will deploy openAI models in Azure openAI resource. 
+
+## Prerequisites
+- Create an [Azure](https://azure.microsoft.com/en-us/features/azure-portal/) account
```
Comment (oshadajay) [inline openapi/azure.openai.deployment/Module.md:7]: Add period.

File: openapi/azure.openai.deployment/Module.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+- Create an [Azure OpenAI resource](https://learn.microsoft.com/en-us/azure/cognitive-services/openai/how-to/create-resource)
```
Comment (oshadajay) [inline openapi/azure.openai.deployment/Module.md:8]: Add period.

File: openapi/azure.openai.deployment/Module.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+- Obtain tokens
+    - Use [this](https://learn.microsoft.com/en-us/azure/cognitive-services/openai/reference#authentication) guide to learn how to generate and use tokens
```
Comment (oshadajay) [inline openapi/azure.openai.deployment/Module.md:10]: Shall we combine L9 and L10? Add period as well.

File: openapi/azure.openai.deployment/Module.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+## Quickstart
+To use the Azure OpenAI Deployment connector in your Ballerina application, update the .bal file as follows:
+
+### Step 1: Import connector
+Import the `ballerinax/azure.openai.deployment` module into the Ballerina project.
+
+    ```ballerina
+
```
Comment (oshadajay) [inline openapi/azure.openai.deployment/Module.md:19]: Remove this line.

File: openapi/azure.openai.deployment/Module.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    import ballerinax/azure.deployment;
+    ```
+
+### Step 2: Create a new connector instance
+Create and initialize a `deployment:Client` with the obtained `apiKey` and a `serviceUrl` from the azure openAI resource
+
+    ```ballerina
+
```
Comment (oshadajay) [inline openapi/azure.openai.deployment/Module.md:27]: Remove this line.

File: openapi/azure.openai.deployment/Module.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    ```ballerina
+    
+    final deployment:Client deploymentClient = check new (
+        apiKeyConfig = {auth: {apiKey: apiKey}},
+        serviceUrl = serviceUrl
+    );
+    ```
+
+### Step 3: Invoke connector operation
+1. Now you can use the operations available within the connector. Note that they are in the form of remote operations.
```
Comment (oshadajay) [inline openapi/azure.openai.deployment/Module.md:35]: Shall we change the style of the note as https://github.com/ballerina-platform/openapi-connectors/pull/754?

File: openapi/azure.openai.deployment/Package.md [hunk 7]
```diff
@@ -0,0 +1,18 @@
+Connects to [Azure OpenAI Deployments API](https://learn.microsoft.com/en-us/rest/api/cognitiveservices/azureopenaistable/deployments/) from Ballerina
```
Comment (oshadajay) [inline openapi/azure.openai.deployment/Package.md:1]: Add period.

File: openapi/azure.openai.deployment/Package.md [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+| Ballerina Language                | 2201.4.1 (Swan Lake Update 4)   | 
+| Azure OpenAI Authoring API        | v2022-12-01-stable              |
+
+## Report issues
+To report bugs, request new features, start new discussions, view project boards, etc., go to the [Ballerina Extended Library repository](https://github.com/ballerina-platform/ballerina-extended-library)
+
+## Useful links
+- Discuss code changes of the Ballerina project in [ballerina-dev@googlegroups.com](mailto:ballerina-dev@googlegroups.com).
+- Chat live with us via our [Discord server](https://discord.gg/ballerinalang).
+- Post all technical questions on Stack Overflow with the [#ballerina](https://stackoverflow.com/questions/tagged/ballerina) tag
```
Comment (oshadajay) [inline openapi/azure.openai.deployment/Package.md:18]: Add period.

File: openapi/azure.openai.deployment/original-openapi.yaml [hunk 9]
```diff
@@ -0,0 +1,1929 @@
+{
```
Comment (oshadajay) [inline openapi/azure.openai.deployment/original-openapi.yaml:1]: Do we have two openapi.yaml specs? I think only the one used with the connector is needed.

Comment (isuruRuhu) [reply openapi/azure.openai.deployment/original-openapi.yaml:1]: That is the practice they have used for connectors with modified yamls, is to have the original yaml too.

File: openapi/azure.openai.deployment/Ballerina.toml [hunk 10]
```diff
@@ -0,0 +1,13 @@
+[package]
+license = ["Apache-2.0"]
+keywords = ["Azure", "OpenAI", "Model", "Deployment"]
+org = "ballerinax"
+name = "azure.openai.deployment"
+icon = "icon.png"
+distribution = "2201.4.1"
+repository = "https://github.com/ballerina-platform/ballerinax-openapi-connectors"
```
[the remark this answers is not in the document]
Comment (isuruRuhu) [reply openapi/azure.openai.deployment/Ballerina.toml:8]: You meant the URL? It is the same for all.

File: openapi/azure.openai.deployment/client.bal [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+        map<string|string[]> httpHeaders = getMapForHeaders(headerValues);
+        DeploymentList response = check self.clientEp->get(resourcePath, httpHeaders);
+        return response;
+    }
+    # Creates a new deployment for the Azure OpenAI resource according to the given specification.
+    #
+    # + apiVersion - The requested API version. 
+    # + payload - The specification of the deployment including the model to deploy and the scale settings. 
+    # + return - The deployment has been successfully created. 
+    resource isolated function post deployments(string apiVersion, Deployment payload) returns Deployment|error {
```
Comment (nadheesh) [inline openapi/azure.openai.deployment/client.bal:77]: we are missing the label tag

File: openapi/azure.openai.deployment/Module.md (hunk 1 above)
Comment (isuruRuhu) [reply openapi/azure.openai.deployment/Module.md:7]: Added the period

File: openapi/azure.openai.deployment/Module.md (hunk 2 above)
Comment (isuruRuhu) [reply openapi/azure.openai.deployment/Module.md:8]: Added the period

File: openapi/azure.openai.deployment/Ballerina.toml (hunk 10 above)
[the remark this answers is not in the document]
Comment (oshadajay) [reply openapi/azure.openai.deployment/Ballerina.toml:8]: The name of the repo is `openapi-connectors` right?

File: openapi/azure.openai.deployment/Module.md [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+    Following is an example on deploying text-davinci-002 model:
+
+    Uploading the data file
```
[the remark this answers is not in the document]
Comment (isuruRuhu) [reply openapi/azure.openai.deployment/Module.md:39]: Removed the line

File: openapi/azure.openai.deployment/Module.md (hunk 5 above)
Comment (isuruRuhu) [reply openapi/azure.openai.deployment/Module.md:27]: Removed the line

File: openapi/azure.openai.deployment/Module.md (hunk 4 above)
Comment (isuruRuhu) [reply openapi/azure.openai.deployment/Module.md:19]: Removed the line

File: openapi/azure.openai.deployment/Module.md (hunk 6 above)
Comment (isuruRuhu) [reply openapi/azure.openai.deployment/Module.md:35]: Modified the line

File: openapi/azure.openai.deployment/Package.md (hunk 7 above)
Comment (isuruRuhu) [reply openapi/azure.openai.deployment/Package.md:1]: Added the period

File: openapi/azure.openai.deployment/Package.md [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+The `azure.openai.deployment` is a [Ballerina](https://ballerina.io/) connector for Azure OpenAI Deployments API.
+
+### Compatibility
+|                                   | Version                         |
+|-----------------------------------|---------------------------------|
+| Ballerina Language                | 2201.4.1 (Swan Lake Update 4)   | 
+| Azure OpenAI Authoring API        | v2022-12-01-stable              |
+
+## Report issues
+To report bugs, request new features, start new discussions, view project boards, etc., go to the [Ballerina Extended Library repository](https://github.com/ballerina-platform/ballerina-extended-library)
```
[the remark this answers is not in the document]
Comment (isuruRuhu) [reply openapi/azure.openai.deployment/Package.md:13]: Added period

File: openapi/azure.openai.deployment/Package.md (hunk 8 above)
Comment (isuruRuhu) [reply openapi/azure.openai.deployment/Package.md:18]: Added the period

File: openapi/azure.openai.deployment/original-openapi.yaml (hunk 9 above)
Comment (oshadajay) [reply openapi/azure.openai.deployment/original-openapi.yaml:1]: Ack

File: openapi/azure.openai.deployment/Ballerina.toml [hunk 14]
```diff
@@ -0,0 +1,13 @@
+[package]
+license = ["Apache-2.0"]
+keywords = ["Azure", "OpenAI", "Model", "Deployment"]
```
Comment (AathmanT) [inline openapi/azure.openai.deployment/Ballerina.toml:3]:
```suggestion
keywords = ["AI/Azure", "Azure OpenAI", "Model", "Deployment"]
```

File: openapi/azure.openai.deployment/Module.md [hunk 15]
```diff
@@ -0,0 +1,49 @@
+## Overview
+This is a generated connector from [Azure OpenAI Deployments API](https://learn.microsoft.com/en-us/rest/api/cognitiveservices/azureopenaistable/deployments/) OpenAPI specification.
+
+The Azure OpenAI Deployments REST API endpoints will deploy openAI models in Azure openAI resource.
```
Comment (AathmanT) [inline openapi/azure.openai.deployment/Module.md:4]:
```suggestion
The Azure OpenAI Deployments REST API endpoints will deploy OpenAI models in Azure OpenAI resource. 
```

File: openapi/azure.openai.deployment/openapi.yaml [hunk 16]
```diff
@@ -0,0 +1,439 @@
...
+  title: Azure OpenAI API version 2022-12-01
+  description: APIs for managing deployments of OpenAI models.
+  x-ballerina-init-description: >
+      The connector initialization requires setting the API credentials.
+      Create an [Azure](https://azure.microsoft.com/en-us/features/azure-portal/) account, an [Azure OpenAI resource](https://learn.microsoft.com/en-us/azure/cognitive-services/openai/how-to/create-resource) and refer [this guide](https://learn.microsoft.com/en-us/azure/cognitive-services/openai/reference#authentication) to learn how to generate and use tokens
+  version: '2022-12-01'
+paths:
+  /deployments:
+    get:
+      tags:
```
Comment (AathmanT) [inline openapi/azure.openai.deployment/openapi.yaml:15]: Add `x-ballerina-display` to all endpoints. Refer
https://github.com/hevayo/ballerina-openai-client/blob/fixes/client_openapi.yaml#L22-L23

File: openapi/azure.openai.deployment/Module.md (hunk 15 above)
Comment (isuruRuhu) [reply openapi/azure.openai.deployment/Module.md]: Modified the module md

File: openapi/azure.openai.deployment/Module.md [hunk 17]
```diff
@@ -0,0 +1,49 @@
+## Overview
+This is a generated connector from [Azure OpenAI Deployments API](https://learn.microsoft.com/en-us/rest/api/cognitiveservices/azureopenaistable/deployments/) OpenAPI specification.
+
+The Azure OpenAI Deployments REST API endpoints will deploy OpenAI models in Azure OpenAI resource. 
+
+## Prerequisites
+- Create an [Azure](https://azure.microsoft.com/en-us/features/azure-portal/) account.
+- Create an [Azure OpenAI resource](https://learn.microsoft.com/en-us/azure/cognitive-services/openai/how-to/create-resource).
+- Obtain the tokens. Refer [this](https://learn.microsoft.com/en-us/azure/cognitive-services/openai/reference#authentication) guide to learn how to generate and use tokens.
```
Comment (anupama-pathirage) [inline openapi/azure.openai.deployment/Module.md:9]:
```suggestion
- Obtain the tokens. Refer to [this](https://learn.microsoft.com/en-us/azure/cognitive-services/openai/reference#authentication) guide to learn how to generate and use tokens.
```

File: openapi/azure.openai.deployment/Module.md [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+    serviceUrl = serviceUrl
+);
+```
+
+### Step 3: Invoke connector operation
+1. Now you can use the operations available within the connector.
+
+>**Note:** that they are in the form of remote operations.
+
+Following is an example on deploying text-davinci-002 model:
```
Comment (anupama-pathirage) [inline openapi/azure.openai.deployment/Module.md:36]:
```suggestion
Following is an example on deploying `text-davinci-002` model:
```

File: openapi/azure.openai.deployment/Module.md (hunk 17 above)
Comment (isuruRuhu) [reply openapi/azure.openai.deployment/Module.md:9]: Modified the PR according to the comment

File: openapi/azure.openai.deployment/Module.md (hunk 18 above)
Comment (isuruRuhu) [reply openapi/azure.openai.deployment/Module.md:36]: Modified the PR according to the comment

**Review by anupama-pathirage — APPROVED** [review]

**Review by NipunaRanasinghe — APPROVED** [review]

**Review by abeykoon — APPROVED** [review]

**Review by abeykoon — APPROVED** [review]

**Pull request merged**